R/build-home-index.R

```diff
@@ -37,19 +43,106 @@ data_home <- function(pkg = ".") {
   ))
 }
 
+default_sidebar_structure <- function() {
+  c("links", "license", "community", "citation", "authors", "dev")
+}
+
 data_home_sidebar <- function(pkg = ".") {
+
   pkg <- as_pkgdown(pkg)
-  if (!is.null(pkg$meta$home$sidebar))
+  if (isFALSE(pkg$meta$home$sidebar))
     return(pkg$meta$home$sidebar)
 
-  paste0(
-    data_home_sidebar_links(pkg),
-    data_home_sidebar_license(pkg),
-    data_home_sidebar_community(pkg),
-    data_home_sidebar_citation(pkg),
-    data_home_sidebar_authors(pkg),
+  html_path <- file.path(pkg$src_path, pkg$meta$home$sidebar$html)
+
+  if (length(html_path)) {
+
+    if (!file.exists(html_path)) {
+      abort(
+        sprintf(
+          "Can't find file %s that was indicated in %s.",
+          pkg$meta$home$sidebar$html,
+          pkgdown_field(pkg, "home", "sidebar", "html")
+          )
+        )
+    }
+    return(paste0(read_lines(html_path), collapse = "\n"))
+  }
+
+  sidebar_structure <- pkg$meta$home$sidebar$structure %||%
+    default_sidebar_structure()
+
+  # compute all default sections
+  sidebar_components <- list(
+    links = data_home_sidebar_links(pkg),
+    license = data_home_sidebar_license(pkg),
+    community = data_home_sidebar_community(pkg),
+    citation = data_home_sidebar_citation(pkg),
+    authors = data_home_sidebar_authors(pkg),
+    dev = sidebar_section("Dev Status", "placeholder")
+  )
+
+  # compute any custom component
+  if (any(!sidebar_structure %in% default_sidebar_structure())) {
```

Comment:
I think you can just unconditionally add any additional components to `sidebar_components` using `modifyList()`, then you check that `setdiff(sidebar_structure, names(sidebar_components)` is empty, and warn about unknown components.